### PR TITLE
UNet class

### PR DESCRIPTION
**Pull request by martinperezmaneiro**
Network with an UNet architecture using ResNet blocks defined in buiding_blocks.py, where you are able to choose its depth and the number of times the input is passed through a basic block in each step of the structure.

### CODE REVIEW

File: next_sparseconvnet/networks/architectures.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+            self.downsample = []
+            self.upsample = []
+            self.basic_down = []
+            self.basic_up = []
+
+            for i in range(level_depth):
+                for j in range(basic_num):
+                    if i != level_depth - 1: #es decir, mientras no llegue al nivel final, hacer 3 basic en ambos lados
+                        self.basic_down.append(ResidualBlock_basic(inplanes, kernel))
```
Comment (mmkekic) [inline next_sparseconvnet/networks/architectures.py]: I think it would be better to somehow separate basic blocks with different size, ie something of a style;
```
self.basic_down = [[] for i in range(level_depth-1)] #initial to  level_depth-1 empty lists
self.bottleneck = [] #lowest layer shared between 2 branches
self.basic_up = [[] for i in range(level_depth-1)] #initial to  level_depth-1 empty lists
```
then loops go as 
```
for i in range(level_depth-1):
    for j in range(basic_num):
        self.basic_down[i].append(ResidualBlock_basic...)
        self.basic_up[i].append(...)
    self.downsample.append(ResidualBlock_downsample...)
    self.upsample.append(....)
    inplanes *=2
#create last layer    
for j in range(basic_num):
    self.bottleneck.append(ResidualBlock_basic...)
    
```

File: next_sparseconvnet/networks/architectures.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                    else: #cuando ya estemos en el último nivel, basta con añadirle 3 basic solo a una de las ramas...
+                        self.basic_down.append(ResidualBlock_basic(inplanes, kernel))
+                if i != level_depth - 1: #lo pongo porque si no me añade en la lista un downsample y un upsample del último nivel (el que considero yo último vamos, en el que solo deberia haber basic)
+                    self.downsample.append(ResidualBlock_downsample(inplanes, kernel, stride)) #lista de los downsamples
+                    self.upsample.append(ResidualBlock_upsample(2 * inplanes, kernel, stride)) #lista de los upsamples, OJO porque está al revés (el primero que debería aplicar es el último de la lista)
+
+                inplanes = inplanes * 2 #en el siguiente nivel tendremos el doble de inplanes...
+
+            self.add = scn.AddTable()
+
```
Comment (mmkekic) [inline next_sparseconvnet/networks/architectures.py]: You are missing output layer, the one that maps the sparse tensor back to dense torch tensor
https://github.com/facebookresearch/SparseConvNet/blob/89818ebd2a508bb05e552168c83d6b60add8a051/sparseconvnet/ioLayers.py#L68

File: next_sparseconvnet/networks/architectures.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        def forward(self, x):
+            x = self.inp(x)
+
+            out_layers = []
+            c = 0 #contador para recorrer los basic
+            for i in range(self.level_depth):
```
Comment (mmkekic) [inline next_sparseconvnet/networks/architectures.py]: Here you can reduce loops and get rid of `c`:
```
tmp_layers = []
#downsample (encoder)
for i in range(self.level_depth):
    for j in range(self.basic_num):
        x = self.basic_down[i][j](x) #we only need i and j to know where the block is in the list
    tmp_layers.append(x)
   
#pass through bottleneck
for i in range(self.basic_num):
    x = self.bottleneck[i](x)
    
#upsample (decoder)
for i in range(self.level_depth-1, 0, -1): #inverse loop
    for j in range(self.basic_num):
        x = self.basic_up[i][j](x) #we only need i and j to know where the block is in the list
        to_add = tmp_layers.pop() #get last item and remove it from the list
        x = self.add([x, to_add])

x = self.output_layer(x)
return x
```

File: next_sparseconvnet/networks/architectures.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+
+import torch
+import sparseconvnet as scn
+from .building_blocks import ResidualBlock_downsample, ResidualBlock_basic, ResidualBlock_upsample
+
+class UNet(torch.nn.Module):
+        def __init__(self, spatial_size, inplanes, kernel, stride, basic_num, level_depth, dim = 3):
+            torch.nn.Module.__init__(self)
```
Comment (mmkekic) [inline next_sparseconvnet/networks/architectures.py]: Is good practice to add docstrings (text inside three quotation marks) where you explain what the class does, input parameters- their types and purpose, and what is the output

**Review by mmkekic — COMMENTED** [review]
Having at least 'it runs' test would be nice before merging this PR

File: next_sparseconvnet/networks/architectures.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+
+import torch
+import sparseconvnet as scn
+from .building_blocks import ResidualBlock_downsample, ResidualBlock_basic, ResidualBlock_upsample
+
+class UNet(torch.nn.Module):
+        def __init__(self, spatial_size, inplanes, kernel, stride, basic_num, level_depth, dim = 3):
```
Comment (mmkekic) [inline next_sparseconvnet/networks/architectures.py]: `kernel` should rather be a list of kernel size per layer, in which case you dont need `level_depth` (it is simply the length of the list)

Comment (mmkekic) [reply next_sparseconvnet/networks/architectures.py]: Also, you need `nclasses` : number of output classes predictions

File: next_sparseconvnet/networks/architectures.py (hunk 2 above)
Comment (mmkekic) [reply next_sparseconvnet/networks/architectures.py]: You also need a 
`nn.LinearLayer(inplanes, nclasses)`
that goes after the `OutputLayer`

**Review by mmkekic — COMMENTED** [review]
To be consistent with general UNet architecture we probably need to pass input tensor through 1-2 Convolutional layers (SubmanifoldConvolution -> BatchNormReLU)
